src/webgldrawer.js

```diff
@@ -0,0 +1,1235 @@
+                    return;
+                }
+
+                // using the context2d pipeline requires a clean rendering (back) buffer to start
+                if(useContext2dPipeline){
+                    // if the rendering buffer has image data currently, write it to the output canvas now and clear it
+
+                    if(renderingBufferHasImageData){
+                        this._outputContext.drawImage(this._renderingCanvas, 0, 0);
+                    }
+
+                    // clear the buffer
+                    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+                    gl.clear(gl.COLOR_BUFFER_BIT); // clear the back buffer
+                }
+
+                // First rendering pass: compose tiles that make up this tiledImage
+                gl.useProgram(this._firstPass.shaderProgram);
+
+                // bind to the framebuffer for render-to-texture if using two-pass rendering, otherwise back buffer (null)
+                if(useTwoPassRendering){
+                    gl.bindFramebuffer(gl.FRAMEBUFFER, this._glFrameBuffer);
+                    // clear the buffer to draw a new image
+                    gl.clear(gl.COLOR_BUFFER_BIT);
+                } else {
+                    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+                    // no need to clear, just draw on top of the existing pixels
+                }
+
+
+
+
+                let overallMatrix = viewMatrix;
+
+                let imageRotation = tiledImage.getRotation(true);
+                // if needed, handle the tiledImage being rotated
+                if( imageRotation % 360 !== 0){
+                    let imageRotationMatrix = Mat3.makeRotation(-imageRotation * Math.PI / 180);
+                    let imageCenter = tiledImage.getBoundsNoRotate(true).getCenter();
+                    let t1 = Mat3.makeTranslation(imageCenter.x, imageCenter.y);
+                    let t2 = Mat3.makeTranslation(-imageCenter.x, -imageCenter.y);
+
+                    // update the view matrix to account for this image's rotation
+                    let localMatrix = t1.multiply(imageRotationMatrix).multiply(t2);
+                    overallMatrix = viewMatrix.multiply(localMatrix);
+                }
+
+                let maxTextures = this._gl.getParameter(this._gl.MAX_TEXTURE_IMAGE_UNITS);
+                let texturePositionArray = new Float32Array(maxTextures * 12); // 6 vertices (2 triangles) x 2 coordinates per vertex
+                let textureDataArray = new Array(maxTextures);
+                let matrixArray = new Array(maxTextures);
+                let opacityArray = new Array(maxTextures);
+
+                // iterate over tiles and add data for each one to the buffers
+                for(let tileIndex = 0; tileIndex < tilesToDraw.length; tileIndex++){
+                    let tile = tilesToDraw[tileIndex].tile;
+                    let index = tileIndex % maxTextures;
+                    let tileContext = tile.getCanvasContext();
```

Comment:
In this location in the code, the canvas associated with each tile is used to look up the corresponding webgl textures in a `Map`. I did it this way because the canvas/context can be shared between multiple tiles if they share image data, and this lets the appropriate texture be used rather than duplicating textures.

Reply:
In my implementation I will not access canvas or image but the tile raw data (usually canvas or image which works all the same), and let the user define custom texture loading logics if they work with custom data.

Reply:
>I did it this way because the canvas/context can be shared between multiple tiles if they share image data, and this lets the appropriate texture be used rather than duplicating textures.

I was not talking about the map, but accessing the texture data. There is no need to access `getCanvasContext` since most protocols are forced to create such canvas for you (or it will be the case in the future), by default the tile data is mostly stored as `Image` (which is not the case for plain image) so calling `getImage()` would be much cheaper in most cases.

Reply:
Have to say that the tile cache logics is under a development too, I have almost finished design of the cache system that gets rid of `context2D` property and removes such bugs.

Reply:
> was not talking about the texture map, but accessing the tile data. There is no need to access `getCanvasContext` since most protocols are forced to create such canvas for you (or it will be the case in the future) on demand when called, by default the tile data is mostly stored as `Image` (which is not the case for plain image) so calling `getImage()` would be much cheaper in most cases.

It isn't important for this to be a canvas and I'd be fine changing to something else. We just need something available on the tile that uniquely maps to the image data that was uploaded to the GPU as a texture.

Reply:
> Btw. getImage() is broken

@Aiosa Yes, this is why I ended up using the context/canvas since that was the one thing that was consistently available. Once the cache refactor goes through this can be revisited.